llumnix/global_scheduler/migration_scheduler.py

```diff
@@ -43,21 +52,58 @@ def __init__(self,
-
+        sorted_src_instance_infos, sorted_dst_instance_infos, pre_migration = self._get_migration_pattern(migrate_target)
+        return self.pair_migration_policy.pair_migration(sorted_src_instance_infos, sorted_dst_instance_infos, pre_migration)
+    def _get_migration_pattern(self, migrate_target:str) -> Dict[str, InstanceInfo]:
+        pre_migration = True
+        if migrate_target == MigrationTarget.GENERAL:
+            # migrate in instances
+            sorted_src_instance_infos = [i for i in reversed(self.sorted_prefill_instance_infos)
+                                if i.num_killed_requests > 0 or i.instance_load_migrate > self.migrate_out_load_threshold]
+            # migrate out instances
+            sorted_dst_instance_infos = [i for i in self.sorted_prefill_instance_infos
+                               if i.num_killed_requests == 0 and i.instance_load_migrate < self.migrate_out_load_threshold]
+        elif migrate_target == MigrationTarget.PREFILL_2_DECODING:
+            sorted_src_instance_infos = [i for i in reversed(self.sorted_prefill_instance_infos)]
+            sorted_dst_instance_infos = [i for i in self.sorted_decoding_instance_infos
+                                            if i.num_killed_requests == 0] # and i.instance_load_migrate < self.migrate_out_load_threshold
+            # TODO[xinyi]: For PDD, add more constaints considering decoding instances load.
+            # if len(sorted_dst_instance_infos) == 0:
+            #     pre_migration = False
+            # else:
+            #     idx = -1
+            #     while len(sorted_src_instance_infos) > len(sorted_dst_instance_infos):
+            #         idx = (idx+1) % len(sorted_src_instance_infos)
+            #         sorted_dst_instance_infos.insert(0, sorted_dst_instance_infos[idx])
+        elif migrate_target == MigrationTarget.DECODING_2_DECODING:
+            sorted_src_instance_infos = [i for i in reversed(self.sorted_decoding_instance_infos)
+                                if i.num_killed_requests > 0 or i.instance_load_migrate > self.migrate_out_load_threshold]
+            sorted_dst_instance_infos = [i for i in self.sorted_decoding_instance_infos
+                               if i.num_killed_requests == 0 and i.instance_load_migrate < self.migrate_out_load_threshold]
+        return sorted_src_instance_infos, sorted_dst_instance_infos, pre_migration
+    
     def update_instance_infos(self,
                               instance_info: Dict[str, InstanceInfo]) -> None:
         self.instance_info = instance_info
 
     def add_instance(self, instance_id: str) -> None:
         self.instance_id_set.add(instance_id)
         self.num_instances = len(self.instance_id_set)
+        if self.available_prefill_instance_num > 0 and len(self.prefill_inst_ids_set) < self.available_prefill_instance_num:
+            self.prefill_inst_ids_set.add(instance_id)
+        else:
+            self.decoding_inst_ids_set.add(instance_id)
```

Comment:
> The current implementation seems to prevent us from designating certain instances as prefill_instances or decode_instances.

For now, since the setting is homogeneous, whether or not to designate has little impact. However, it is still necessary to designate certain instances as prefill_instances or decode_instances, which will be the next step.